lib/api.js

```diff
@@ -0,0 +1,153 @@
+import base from 'taskcluster-base';
+import common from './common';
+import slugid from 'slugid';
+import _ from 'lodash';
+
+/**
+ * Check to see if a resource is stale
+ **/
+function hasExpired(expires) {
+  let currentTime = new Date();
+  let expireTime = new Date(expires);
+  if (currentTime > expireTime) {
+    return true;
+  }
+  return false;
+};
+
+/** API end-point for version v1/
+ *
+ * In this API implementation we shall assume the following context:
+ * {
+ *   publisher:      // publisher from base.Exchanges
+ * }
+ */
+var api = new base.API({
+  title:        "TaskCluster Secrets API Documentation",
+  description: [
+    "The secrets service, typically available at",
+    "`tools.taskcluster.net`, is responsible for managing",
+    "secure data in TaskCluster."
+  ].join('\n')
+});
+
+// Export API
+module.exports = api;
+
+/** Define tasks */
+api.declare({
+  method:      'put',
+  route:       '/name/:name(*)',
+  deferAuth:   true,
+  name:        'set',
+  input:       common.SCHEMA_PREFIX_CONST + "secret.json#",
+  scopes:      [['secrets:set:<name>']],
+  title:       'Create New Secret',
+  description: 'Set a secret associated with some key.'
+}, async function(req, res) {
+    req.satisfies(req.params);
+    let {name} = req.params;
+    let {secret, expires} = req.body;
+    try {
+      await this.entity.create({
+        name:       name,
+        secret:     secret,
+        expires:    new Date(expires)
+      });
+    } catch(e) {
+      res.status(e.statusCode).send(e.code);
```

Comment:
here you need to handle `EntityAlreadyExists` with care.. Proposal for catch block:
```js
if (e.code !== 'EntityAlreadyExists') {
  throw e;
}
let secret = await this.entity.load({name});
if (!_.isEqual(secret.secret, secret) || (new Date(expires)).getTime() !== secret.expires.getTime()) {
  return res.status(409).json({message: "A different secret with the same name exists!"});
}
```
This is similar to what queue does here:
https://github.com/taskcluster/taskcluster-queue/blob/master/routes/v1.js#L299-L344
In fact all of our HTTP end-points tries to do exactly this. Because it ensures that the request is successful, even if getting the response failed and we had to retry, but the operation was applied. (keyword: idempotent).

Forwarding unknown errors is to end-users is bad idea as we can't tell what is in the error message.
It's probably not a problem, but if you just let the error raise up to the API level, the API wrapper will log the error, any properties on the error and generate a incidentId, then print incidentId to log and return a 500 error telling user to what the incidentId is... This way you can debug it when a user tells you that something triggers a 500 error.